### PR TITLE
lots of goodies

### PR DESCRIPTION
**Pull request by amnona**
- interactive html export of heatmap
- faster sum/mean for filtering
- move analysis into database modules

### CODE REVIEW

File: calour/filtering.py [hunk 1]
```diff
@@ -164,31 +165,39 @@ def filter_by_data(exp, predicate, axis=0, negate=False, inplace=False, **kwargs
...
+                 'mean_abundance': _mean_abundance}
+    if predicate in fast_func:
+        logger.debug('filter_by_data using fast function %r' % predicate)
+        select = fast_func[predicate](exp.data, axis=1-axis, **kwargs)
     else:
-        select = np.apply_along_axis(predicate, 1 - axis, exp.data, **kwargs)
+        func = {'sum_abundance': _sum_abundance,
+                'freq_ratio': _freq_ratio,
+                'unique_cut': _unique_cut,
+                'mean_abundance': _mean_abundance,
```
Comment (RNAer) [inline calour/filtering.py]: delete `sum_abundance` and `mean_abundance`

File: calour/transforming.py [hunk 2]
```diff
@@ -251,3 +251,28 @@ def normalize_by_subset_features(exp, features, total=10000, negate=True, inplac
         newexp = deepcopy(exp)
     newexp.data = total * data / use_reads[:, None]
     return newexp
+
+
+def normalize_compositional(exp, min_frac=0.05, total=10000, inplace=False):
+    '''Normalize each sample by ignoring the features with mean>=min_frac in all the experiment
```
Comment (RNAer) [inline calour/transforming.py:257]: add an blanke line

File: calour/transforming.py [hunk 3]
```diff
@@ -251,3 +251,28 @@ def normalize_by_subset_features(exp, features, total=10000, negate=True, inplac
...
+        False (default) to create a new experiment, True to normalize in place
+
+    Returns
+    -------
+    ``Experiment``
+        The normalized experiment. Note that all features are normalized (including the ones with mean>=min_frac)
+    '''
+    comp_features = exp.filter_mean(min_frac)
+    logger.info('ignoring %d features' % comp_features.shape[1])
+    newexp = exp.normalize_by_subset_features(comp_features.feature_metadata.index.values, total=total, negate=True, inplace=inplace)
```
Comment (RNAer) [inline calour/transforming.py]: line too long!

File: calour/heatmap/plotgui_qt5.py [hunk 4]
```diff
@@ -338,32 +337,32 @@ def save_fasta(self):
         self.gui.exp.save_fasta(str(filename), seqs)
 
     def enrichment(self):
+        exp = self.gui.exp
```
Comment (RNAer) [inline calour/heatmap/plotgui_qt5.py]: document this func

File: calour/heatmap/plotgui_qt5.py [hunk 5]
```diff
@@ -338,32 +337,32 @@ def save_fasta(self):
...
         group1_seqs = self.gui.get_selected_seqs()
-        allseqs = self.gui.exp.feature_metadata.index.values
+        allseqs = exp.feature_metadata.index.values
         group2_seqs = list(set(allseqs).difference(set(group1_seqs)))
```
Comment (RNAer) [inline calour/heatmap/plotgui_qt5.py]: use `set1 - set2`, which is more clear

File: calour/export_html.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+import numpy as np
+
+from .transforming import log_n
+from .heatmap.heatmap import _transition_index
+
+logger = getLogger(__name__)
+
+
+def _list_to_string(l):
+    '''Convert a list to a string representation of the list
```
Comment (RNAer) [inline calour/export_html.py:23]: why not just `repr(a_list)`?

File: calour/export_html.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        html_page = html_page.replace('// xtick_labels go here', 'xtick_labels=%s' %
+                                      _list_to_string(xticklabels))
+
+    # embed the figure png into the html page
+    figfile = BytesIO()
+    fig.savefig(figfile, format='png', dpi=300)
+    figfile.seek(0)  # rewind to beginning of file
+    import base64
+    figdata_png = base64.b64encode(figfile.getvalue())
+    figfile.close()
```
Comment (RNAer) [inline calour/export_html.py]: use `with`

File: calour/export_html.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    html_page = html_page.replace('**image_goes_here**', urllib.parse.quote(figdata_png))
```
Comment (RNAer) [inline calour/export_html.py]: mv `urllib.parse.quote` to `with` block. just use the final value in this `replace` call

**Review by RNAer — CHANGES_REQUESTED** [review]

File: calour/filtering.py (hunk 1 above)
Comment (amnona) [reply calour/filtering.py]: yes :)

File: calour/transforming.py (hunk 2 above)
Comment (amnona) [reply calour/transforming.py:257]: done

File: calour/transforming.py (hunk 3 above)
Comment (amnona) [reply calour/transforming.py]: ok

File: calour/heatmap/plotgui_qt5.py (hunk 4 above)
Comment (amnona) [reply calour/heatmap/plotgui_qt5.py]: done

File: calour/heatmap/plotgui_qt5.py (hunk 5 above)
Comment (amnona) [reply calour/heatmap/plotgui_qt5.py]: ok

File: calour/export_html.py (hunk 6 above)
Comment (amnona) [reply calour/export_html.py:23]: never

File: calour/export_html.py (hunk 7 above)
Comment (amnona) [reply calour/export_html.py]: done by you :)

File: calour/export_html.py (hunk 8 above)
Comment (amnona) [reply calour/export_html.py]: done by you :)

File: calour/heatmap/plotgui_qt5.py [hunk 9]
```diff
@@ -337,10 +337,17 @@ def save_fasta(self):
...
+        Iterate over all databases that support enrichment analysis. For each such database,
+        get the list of enriched terms in the selected set of features (compared to the other features
+        in the experiment). Then display the list of these terms in a new qt5 window with blue terms
+        for ones enriched in the selected group, red terms for ones enriched in the unselected set of features
+        '''
         exp = self.gui.exp
         group1_seqs = self.gui.get_selected_seqs()
         allseqs = exp.feature_metadata.index.values
-        group2_seqs = list(set(allseqs).difference(set(group1_seqs)))
+        group2_seqs = list(set(allseqs) - set(group1_seqs))
```
Comment (RNAer) [inline calour/heatmap/plotgui_qt5.py]: no need to convert to list

**Review by RNAer — CHANGES_REQUESTED** [review]

File: calour/heatmap/plotgui_qt5.py (hunk 9 above)
Comment (amnona) [reply calour/heatmap/plotgui_qt5.py]: true :)

**Pull request merged**